source/Halibut/Transport/SecureListeningClient.cs

```diff
@@ -0,0 +1,157 @@
+
+        public SecureListeningClient(ServiceEndPoint serviceEndpoint, X509Certificate2 clientCertificate, ILog log, ConnectionManager connectionManager)
+        {
+            this.ServiceEndpoint = serviceEndpoint;
+            this.clientCertificate = clientCertificate;
+            this.log = log;
+            this.connectionManager = connectionManager;
+        }
+
+        public ServiceEndPoint ServiceEndpoint { get; }
+
+        public void ExecuteTransaction(Action<MessageExchangeProtocol> protocolHandler)
+        {
+            var retryInterval = ServiceEndpoint.RetryListeningSleepInterval;
+
+            Exception lastError = null;
+
+            // retryAllowed is also used to indicate if the error occurred before or after the connection was made
+            var retryAllowed = true;
+            var watch = Stopwatch.StartNew();
+            for (var i = 0; i < ServiceEndpoint.RetryCountLimit && retryAllowed && watch.Elapsed < ServiceEndpoint.ConnectionErrorRetryTimeout; i++)
+            {
+                if (i > 0) log.Write(EventType.Error, "Retry attempt {0}", i);
+
+                try
+                {
+                    lastError = null;
+
+                    IConnection connection = null;
+                    try
+                    {
+                        connection = connectionManager.AcquireConnection(new TcpConnectionFactory(clientCertificate), ServiceEndpoint, log);
+
+                        // Beyond this point, we have no way to be certain that the server hasn't tried to process a request; therefore, we can't retry after this point
+                        retryAllowed = false;
+
+                        protocolHandler(connection.Protocol);
+                    }
+                    catch
+                    {
+                        connection?.Dispose();
+                        if (connectionManager.IsDisposed)
+                            return;
+                        throw;
+                    }
+
+                    // Only return the connection to the pool if all went well
+                    connectionManager.ReleaseConnection(ServiceEndpoint, connection);
+                }
+                catch (AuthenticationException aex)
+                {
+                    log.WriteException(EventType.Error, $"Authentication failed while setting up connection to {(ServiceEndpoint == null ? "(Null EndPoint)" : ServiceEndpoint.BaseUri.ToString())}", aex);
+                    lastError = aex;
+                    retryAllowed = false;
+                    break;
+                }
+                catch (SocketException cex) when (cex.SocketErrorCode == SocketError.ConnectionRefused)
+                {
+                    log.Write(EventType.Error, $"The remote host at {(ServiceEndpoint == null ? "(Null EndPoint)" : ServiceEndpoint.BaseUri.ToString())} refused the connection, this may mean that the expected listening service is not running.");
```

Comment:
nothing major. just a tidy thing, the repeated `(ServiceEndpoint == null ? "(Null EndPoint)" : ServiceEndpoint.BaseUri.ToString()` could be put into a local method.